lightning/src/ln/channelmanager.rs

```diff
@@ -99,6 +99,7 @@ enum PendingHTLCRouting {
 	Receive {
 		payment_data: msgs::FinalOnionHopData,
 		incoming_cltv_expiry: u32, // Used to track when we should expire pending HTLCs that go unclaimed
+		keysend_preimage: Option<PaymentPreimage>,
```

Comment:
I feel like this commits the cardinal sin of not having the datastructure exactly match the codepaths, and it shows - we have to adapt the receive stuff to create a fake inbound payment request to make it work, plus add a fake payment_secret, which is not at all fun. I think, in general, we should always, always prefer to lean on the typesystem and add more enum variants so that every use of the object is really clear what's going on and what the conditions we can be in are.

Concretely, what do you think about this changeset? I didn't implement the actual receive-side logic, but I don't think it would be toooo hard, and its much clearer what's going on there. https://git.bitcoin.ninja/index.cgi?p=rust-lightning;a=blobdiff;f=lightning/src/ln/channelmanager.rs;h=1536ddf085e5ca55cd7acb922fe89dd9bfc78f13;hp=f5f75b50c75fef5f2cdbb501f0b0971a806de844;hb=0bcb7a38bb46e9abe479a7dbc3d08ccfa633530b;hpb=de9d7f7b2bea717cadfb1ea1fc6de3622d6c8751